src/libraries/System.Private.CoreLib/src/System/IO/DerivedFileStreamStrategy.cs

```diff
@@ -0,0 +1,141 @@
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    // this type exists so we can avoid GetType() != typeof(FileStream) checks in FileStream
+    // when FileStream was supposed to call base.Method() for such cases, we just call _fileStream.BaseMethod()
+    // for everything else we fall back to the actual strategy (like FileStream does)
+    //
+    // it's crucial to NOT use the "base" keyoword here! everything must be using _fileStream or _strategy
+    internal sealed class DerivedFileStreamStrategy : FileStreamStrategy
+    {
+        private readonly FileStreamStrategy _strategy;
+
+        internal DerivedFileStreamStrategy(FileStream fileStream, FileStreamStrategy strategy) : base(fileStream) => _strategy = strategy;
+
+        public override bool CanRead => _strategy.CanRead;
+
+        public override bool CanWrite => _strategy.CanWrite;
+
+        public override bool CanSeek => _strategy.CanSeek;
+
+        public override long Length => _strategy.Length;
+
+        public override long Position
+        {
+            get => _strategy.Position;
+            set => _strategy.Position = value;
+        }
+
+        internal override bool IsAsync => _strategy.IsAsync;
+
+        internal override string Name => _strategy.Name;
+
+        internal override SafeFileHandle SafeFileHandle => _strategy.SafeFileHandle;
+
+        internal override bool IsClosed => _strategy.IsClosed;
+
+        internal override void Lock(long position, long length) => _strategy.Lock(position, length);
+
+        internal override void Unlock(long position, long length) => _strategy.Unlock(position, length);
+
+        public override long Seek(long offset, SeekOrigin origin) => _strategy.Seek(offset, origin);
+
+        public override void SetLength(long value) => _strategy.SetLength(value);
+
+        public override int ReadByte() => _strategy.ReadByte();
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+            => _strategy.IsAsync
+                ? _strategy.BeginRead(buffer, offset, count, callback, state)
+                : _fileStream.BaseBeginRead(buffer, offset, count, callback, state);
+
+        public override int EndRead(IAsyncResult asyncResult)
+            => _strategy.IsAsync ? _strategy.EndRead(asyncResult) : _fileStream.BaseEndRead(asyncResult);
+
+        public override int Read(byte[] buffer, int offset, int count) => _strategy.Read(buffer, offset, count);
+
+        // If this is a derived type, it may have overridden Read(byte[], int, int) prior to this Read(Span<byte>)
+        // overload being introduced.  In that case, this Read(Span<byte>) overload should use the behavior
+        // of Read(byte[],int,int) overload.
+        public override int Read(Span<byte> buffer)
+            => _fileStream.BaseRead(buffer);
+
+        // If we have been inherited into a subclass, the Strategy implementation could be incorrect
```

Comment:
I would prefer not to use `FileStreamStrategyBase` as it suggest that it has to be something that derives from this type, while it could be a brand new type like `IoUringFileStreamStrategy` that might not derive from `FileStreamStrategyBase`, just from `FileStreamStrategy`